File: src/main/java/com/hileco/cortex/analysis/VisualGraph.java
package com.hileco.cortex.analysis;

import com.hileco.cortex.analysis.edges.EdgeFlowMapping;
import guru.nidi.graphviz.attribute.Color;
import guru.nidi.graphviz.attribute.RankDir;
import guru.nidi.graphviz.attribute.Records;
import guru.nidi.graphviz.engine.Format;
import guru.nidi.graphviz.engine.Graphviz;
import guru.nidi.graphviz.model.Compass;
import guru.nidi.graphviz.model.Link;
import guru.nidi.graphviz.model.Node;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import static guru.nidi.graphviz.attribute.Records.rec;
import static guru.nidi.graphviz.model.Factory.graph;
import static guru.nidi.graphviz.model.Factory.node;
import static guru.nidi.graphviz.model.Factory.port;
import static guru.nidi.graphviz.model.Link.between;

public class VisualGraph {

    private final Map<Integer, Node> vizNodeMapping;
    private guru.nidi.graphviz.model.Graph vizGraph;

    public VisualGraph() {
        this.vizNodeMapping = new HashMap<>();
        this.vizGraph = graph()
                .directed()
                .graphAttr().with(RankDir.LEFT_TO_RIGHT);
    }

    private void map(GraphBlock graphBlock) {
        var records = new ArrayList<String>();
        var lines = new ArrayList<Integer>();
        for (var graphNode : graphBlock.getGraphNodes()) {
            lines.add(graphNode.getLine());
            records.add(rec(graphNode.getLine().toString(), String.format("%d: %s", graphNode.getLine(), graphNode.getInstruction().get().toString())));
        }
        var node = node(lines.iterator().next().toString())
                .with(Records.of(records.toArray(new String[0])))
                .with(Color.WHITE.fill());
        lines.forEach(line -> this.vizNodeMapping.put(line, node));
        this.vizGraph = this.vizGraph.with(node);
    }

    private void map(EdgeFlowMapping edgeFlowMapping) {
        var nodeMapping = new HashMap<Node, ArrayList<Link>>();
        edgeFlowMapping.getFlowsFromSource().forEach((source, flows) -> {
            if (source != null) {
                var sourceVizNode = this.vizNodeMapping.get(source);
                flows.forEach(flow -> {
                    if (flow.getTarget() != null) {
                        var vizLinkSources = nodeMapping.computeIfAbsent(sourceVizNode, ignored -> new ArrayList<>());
                        var targetVizNode = this.vizNodeMapping.get(flow.getTarget());
                        if (!sourceVizNode.equals(targetVizNode)) {
                            vizLinkSources.add(between(port(source.toString()), targetVizNode.port(flow.getTarget().toString(), Compass.WEST)));
                        }
                    }
                });
            }
        });
        nodeMapping.forEach((node, links) -> this.vizGraph = this.vizGraph.with(node.link(links.toArray(new Link[0]))));
    }

    public void map(Graph graph) {
        graph.getGraphBlocks().forEach(this::map);
        graph.getEdges().stream().flatMap(EdgeFlowMapping.UTIL::filter).forEach(this::map);
    }

    public byte[] toBytes() throws IOException {
        var graphViz = Graphviz.fromGraph(this.vizGraph);
        var outputStream = new ByteArrayOutputStream();
        graphViz.render(Format.PNG).toOutputStream(outputStream);
        return outputStream.toByteArray();
    }
}

File: src/main/java/com/hileco/cortex/analysis/VisualGraph.kt
package com.hileco.cortex.analysis

import com.hileco.cortex.analysis.edges.FlowMapping
import guru.nidi.graphviz.attribute.Color
import guru.nidi.graphviz.attribute.RankDir
import guru.nidi.graphviz.attribute.Records
import guru.nidi.graphviz.attribute.Records.rec
import guru.nidi.graphviz.engine.Format
import guru.nidi.graphviz.engine.Graphviz
import guru.nidi.graphviz.model.Compass
import guru.nidi.graphviz.model.Factory.*
import guru.nidi.graphviz.model.Link
import guru.nidi.graphviz.model.Link.between
import guru.nidi.graphviz.model.Node
import java.io.ByteArrayOutputStream
import java.util.*

class VisualGraph {
    private val vizNodeMapping: MutableMap<Int, Node> = HashMap()
    private var vizGraph: guru.nidi.graphviz.model.Graph = graph().directed().graphAttr().with(RankDir.LEFT_TO_RIGHT)

    private fun map(graphBlock: GraphBlock) {
        val records = ArrayList<String>()
        val lines = ArrayList<Int>()
        for (graphNode in graphBlock.graphNodes) {
            lines.add(graphNode.line)
            records.add(rec(graphNode.line.toString(), "${graphNode.line}: ${graphNode.instruction}"))
        }
        val node = node("${lines.first()}")
                .with(Records.of(*records.toTypedArray()))
                .with(Color.WHITE.fill())
        lines.forEach { line -> vizNodeMapping[line] = node }
        vizGraph = vizGraph.with(node)
    }

    private fun map(flowMapping: FlowMapping) {
        val nodeMapping = HashMap<Node, ArrayList<Link>>()
        flowMapping.flowsFromSource.forEach { source, flows ->
            if (source != null) {
                val sourceVizNode = vizNodeMapping[source]!!
                flows.forEach { flow ->
                    if (flow.target != null && flow.type.jumps) {
                        val vizLinkSources = nodeMapping.computeIfAbsent(sourceVizNode) { ArrayList() }
                        val targetVizNode = vizNodeMapping[flow.target]!!
                        vizLinkSources.add(between(port("$source"), targetVizNode.port(flow.target.toString(), Compass.WEST)))
                    }
                }
            }
        }
        nodeMapping.forEach { node, links -> vizGraph = vizGraph.with(node.link(*links.toTypedArray())) }
    }

    fun map(graph: Graph) {
        graph.graphBlocks.forEach { this.map(it) }
        graph.edgeMapping.get(FlowMapping::class.java).forEach {
            this.map(it)
        }
    }

    fun toBytes(): ByteArray {
        val graphViz = Graphviz.fromGraph(vizGraph)
        val outputStream = ByteArrayOutputStream()
        graphViz.render(Format.PNG).toOutputStream(outputStream)
        return outputStream.toByteArray()
    }
}
